src/vcpkg/base/system.cpp

```diff
@@ -417,17 +417,33 @@ namespace vcpkg
         return ProgramW6432;
     }
 
-    int get_concurrency()
+    unsigned int get_concurrency()
     {
-        static int concurrency = [] {
+        static unsigned int concurrency = [] {
             auto user_defined_concurrency = get_environment_variable("VCPKG_MAX_CONCURRENCY");
             if (user_defined_concurrency)
             {
-                return std::stoi(user_defined_concurrency.value_or_exit(VCPKG_LINE_INFO));
+                try
+                {
+                    int res = std::stoi(user_defined_concurrency.value_or_exit(VCPKG_LINE_INFO));
+                    if (!(res > 0))
+                    {
+                        Checks::msg_exit_with_message(VCPKG_LINE_INFO,
+                                                      msgEnvInvalidMaxConcurrency,
+                                                      msg::env_var = "VCPKG_MAX_CONCURRENCY",
+                                                      msg::value = res);
+                    }
+                    return static_cast<unsigned int>(res);
+                }
+                catch (std::exception&)
+                {
+                    Checks::msg_exit_with_message(
+                        VCPKG_LINE_INFO, msgOptionMustBeInteger, msg::option = "VCPKG_MAX_CONCURRENCY");
+                }
```

Comment:
We should minimize the coverage of the `try` to ensure we don't capture any unexpected exceptions.

```suggestion
                int res;
                try
                {
                    res = std::stoi(user_defined_concurrency.value_or_exit(VCPKG_LINE_INFO));
                }
                catch (std::exception&)
                {
                    Checks::msg_exit_with_message(
                        VCPKG_LINE_INFO, msgOptionMustBeInteger, msg::option = "VCPKG_MAX_CONCURRENCY");
                }
                if (!(res > 0))
                {
                    Checks::msg_exit_with_message(VCPKG_LINE_INFO,
                                                  msgEnvInvalidMaxConcurrency,
                                                  msg::env_var = "VCPKG_MAX_CONCURRENCY",
                                                  msg::value = res);
                }
                return static_cast<unsigned int>(res);
            }
```